include/nccl_ofi_param.h

```diff
@@ -16,54 +16,61 @@ extern "C" {
 #define OFI_NCCL_PARAM_INT(name, env, default_value) \
 pthread_mutex_t ofi_nccl_param_lock_##name = PTHREAD_MUTEX_INITIALIZER; \
 int64_t ofi_nccl_##name() { \
-    assert(default_value != -1LL); \
-    static int64_t value = -1LL; \
+    static bool initialized = false; \
+    static int64_t value = default_value; \
+    if (initialized) { \
+	return value; \
+    } \
     pthread_mutex_lock(&ofi_nccl_param_lock_##name); \
     int64_t v; \
-    char *str; \
-    if (value == -1LL) { \
-        value = default_value; \
+    char *str, *endptr; \
+    if (!initialized) { \
         str = getenv("OFI_NCCL_" env); \
         if (str && strlen(str) > 0) { \
             errno = 0; \
-            v = strtoll(str, NULL, 0); \
-            if (!errno) { \
+            v = strtoll(str, &endptr, 0); \
+	    printf("errno: %d, str: %p endptr: %p\n", errno, str, endptr); \
+            if (errno || str == endptr || *endptr != '\0') { \
+                NCCL_OFI_INFO(NCCL_INIT | NCCL_NET, \
+                    "Invalid value %s provided for %s environment variable, using default %lu", \
+                    str, "OFI_NCCL_" env, value); \
+            } else { \
                 value = v; \
                 NCCL_OFI_INFO(NCCL_INIT | NCCL_NET, "Setting %s environment variable to %lu", \
                               "OFI_NCCL_" env, value); \
-            } else { \
-                NCCL_OFI_INFO(NCCL_INIT | NCCL_NET, \
-                    "Invalid value %s provided for %s environment variable, using default %s", \
-                    str, "OFI_NCCL_" env, value); \
             } \
         } \
+	initialized = true; \
     } \
     pthread_mutex_unlock(&ofi_nccl_param_lock_##name); \
     return value; \
 }
 
 #define OFI_NCCL_PARAM_STR(name, env, default_value) \
 pthread_mutex_t ofi_nccl_param_lock_##name = PTHREAD_MUTEX_INITIALIZER; \
-char *ofi_nccl_##name() { \
-    assert(default_value != NULL); \
-    static char *value = NULL; \
+const char *ofi_nccl_##name() { \
+    static bool initialized = false; \
+    static const char *value = default_value; \
+    if (initialized) { \
+	return value; \
+    } \
     pthread_mutex_lock(&ofi_nccl_param_lock_##name); \
     char *str; \
-    if (value == NULL) { \
-        value = strdup(default_value); \
+    if (!initialized) { \
         str = getenv("OFI_NCCL_" env); \
-        if (str && strlen(str) > 0) { \
-            errno = 0; \
+        if (str) { \
             value = strdup(str); \
```

Comment:
Now that this function returns const char*, is there any reason to use strdup? The string returned by getenv should live for the duration of the application, right?

Reply:
Yes.  Because `getenv()`'s standard definition is evil.  From the Linux `getenv()` man page:

> The  implementation  of  getenv() is not required to be reentrant.  The string pointed to by the return value of getenv() may be statically allocated, and  can  be  modified  by  a  subsequent  call  to  getenv(),  putenv(3), setenv(3), or unsetenv(3).

Now, no modern implementation I'm aware of does anything so stupid, but technically, you shouldn't store pointers to the return of `getenv()` for any real period of time.